### ISSUE TITLE
`zca_whitening` causes memory leak in `ImageDataGenerator.fit`

### ISSUE THREAD

**Issue by wottpal** [keras-team/keras-preprocessing#55]
* Keras `2.2.2`
* TensorFlow `1.10.1`
* Python `3.6.6`
* macOS `10.13.6`

I'm trying to generate augmentations of my training data with `zca_whitening` and an `ImageDataGenerator`. But when I try to fit the generator (which is mandatory when using `zca_whitening`) the python process eats more and more memory (100Gb+) until it gets killed by the system.

This small example can cause the leak:

```
import numpy as np
from keras.preprocessing.image import ImageDataGenerator

def cause_leak():
    idg = ImageDataGenerator(zca_whitening = True)
    random_sample = np.random.random((1, 250, 250, 3))
    idg.fit(random_sample)

cause_leak()
```

The terminal output only consists of a warning saying that `featurewise_center` is overwritten when enabling `zca_whitening`. I don't think this is related to the problem but who knows.

Does anybody know a workaround?

**Comment by wottpal** [comment]
Previously an issue in the main repo: 
https://github.com/keras-team/keras/issues/11058

**Comment by Dref360** [comment]
This is not a memory leak. We would call this a memory leak if multiple calls to `fit` would cause a failure because of unreleased resources.

Computing the SVD of a matrix with 187500 elements is memory intensive.
You cannot use `zca` in your case.

**Comment by wottpal** [comment]
Do you have some tips on how I would use `zca_whitening` then? I mean 187.500 sounds like a large number but it‘s only a single 250x250 image 😔 And in a real world scenario I wouldn‘t fit the `ImageDataGenerator` on a single image but multiple hundreds-thousands I guess. Some guidance would be great!

**Comment by csheaff** [comment]
Hello, I am having the same issue using a single 384 x 384 image.

**Comment by rragundez** [comment]
This is known problem with the ZCA method. There is no real immediate way around it. The problem is with the calculation of the sigma matrix which is a dot product. https://github.com/keras-team/keras-preprocessing/blob/master/keras_preprocessing/image/image_data_generator.py#L938

This issue should be closed. @Dref360

**Comment by Nestak2** [comment]
I agree that the problem is that the zca-calculation takes very, very long, because it [uses][2] `linalg.svd()` which is computationally heavy even on matrices (n*m*3), n=m=100. There are some approaches to calculate a quick approximation of the svd, like randomized svd by [scikitlearn][3], truncated svd by [irlb][4], Lanczos method svd, but they are not always numerically stable.

I think I have found another very simple approximation, that appears to give good enough results and is very quick! It helps in case you have a data matrix with shape (m x n), where m is much smaller than n - e.g. you have much less images (m=1000) than number of pixels (n=100 x 100 x 3 = 30000 pixel). In such a case keras would calculate the `linalg.svd(sigma)`, with `sigma.shape=(30000,30000)`, which is computationally too exhausting and takes forever. But in a good approximation, instead of calculating the svd on a (n x n) matrix you could calculate it on a (m x m) matrix, by just rotating the input data X or flipping the order of the `sigma` calculation to look like `sigma = np.dot(flat_x, flat_x.T) / flat_x.shape[0]`. With this approach the calculation takes only ca. 10 sec, if m=1000 and n=30000. The good thing is that the eigenvectors of the `linalg.svd(sigma)` are the same in both cases up to a factor, see [here][5], slide 30. You could test this on a dataset of yours or on the inbuilt cifer set with `from keras.datasets import cifar10`.

Unfortunately, I don't have the mathematical argumentation of this approach, but I have done visual inspections of the zca-images from both calculations and they look very similar: 

[![enter image description here][6]][6]
[![enter image description here][7]][7]

And below is the modified keras code for creating the zca calculations. You could use it to make modifications into your keras zca code, located [here][2]: There are only 2 modifications I have made that get activated if `zca_rotated=True`:

    from keras.datasets import cifar10
    import numpy as np
    from scipy import linalg
    
    (X_train, y_train), (X_test, y_test) = cifar10.load_data()
    X = X_train[:1000]
    flat_x = np.reshape(x, (x.shape[0], x.shape[1] * x.shape[2] * x.shape[3]))
    
    # line below lets you choose if you want to calculate the standard zca, like in keras (=False)
    # or if you want to treat the input data X as rotated to simplify the calculation if X.shape[0]=m < X.shape[1]=n
    zca_rotated = True
    
    # normalize x:
    flat_x = flat_x / 255.
    flat_x = flat_x - flat_x.mean(axis=0)
    
    # 1ST CHANGE IN CALCULATION BELOW
    if zca_rotated == False:
        sigma = np.dot(flat_x.T, flat_x) / flat_x.shape[0]
    elif zca_rotated == True:
        sigma = np.dot(flat_x, flat_x.T) / flat_x.shape[0]
    
    u, s, _ = linalg.svd(sigma)
    s_inv = 1. / np.sqrt(s[np.newaxis] + 0.1) # the 0.1 is the epsilon value for zca
    principal_components = (u * s_inv).dot(u.T)
    
    # below is the calculation of the actual zca-image with the fitted parameters from above
    # 2ND CHANGE IN LINE BELOW
    if zca_rotated == False:
        whitex = np.dot(flat_x, principal_components)
    elif zca_rotated == True:
        whitex = np.dot(flat_x.T, principal_components)


  [1]: https://github.com/keras-team/keras-preprocessing/issues/55
  [2]: https://github.com/keras-team/keras-preprocessing/blob/master/keras_preprocessing/image/image_data_generator.py#L938
  [3]: https://scikit-learn.org/stable/modules/generated/sklearn.utils.extmath.randomized_svd.html
  [4]: http://bwlewis.github.io/irlba/
  [5]: http://www.cs.cmu.edu/~aarti/Class/10701_Spring14/slides/PCA.pdf
  [6]: https://i.stack.imgur.com/AUdeD.png
  [7]: https://i.stack.imgur.com/1rBca.png
